### PR TITLE
vstart: generate environment script suitable for sourcing

### PR DESCRIPTION
**Pull request by jtlayton**
It took me a bit of effort to be able to run cephfs-shell against a vstart cluster without installing all of the python modules. This just has vstart generate a shell script that sets up the environment properly that can be sourced into the shell. Also a trivial patch to change the mode of cephfs-shell in the tree to be executable, which makes it easier to run out of tree.

### CODE REVIEW

**Comment by jtlayton** [comment]
retest this please

File: src/vstart.sh [hunk 1]
```diff
@@ -1232,5 +1232,17 @@ if [ "$CEPH_DIR" != "$PWD" ]; then
...
 fi
 
+# convenience wrapper for running cephfs-shell out of build dir
+if [ $MDS -gt 0  -a -n "$CEPH_ROOT" ]; then
+	cat > $CEPH_BIN/vstart-cephfs-shell << EOF
+#!/bin/bash
+
+export LD_LIBRARY_PATH=$CEPH_LIB
+export PYTHONPATH=$PYBIND:$CEPH_LIB/cython_modules/lib.$CEPH_PY_VERSION_MAJOR
+exec $CEPH_ROOT/src/tools/cephfs/cephfs-shell
```
Comment (batrick) [inline src/vstart.sh]: Is the `cephfs-shell` not installed in `$CEPH_BIN`? Shouldn't it refer to that?

File: src/vstart.sh [hunk 2]
```diff
@@ -1232,5 +1232,17 @@ if [ "$CEPH_DIR" != "$PWD" ]; then
     echo "export CEPH_KEYRING=$keyring_fn"
 fi
 
+# convenience wrapper for running cephfs-shell out of build dir
+if [ $MDS -gt 0  -a -n "$CEPH_ROOT" ]; then
+	cat > $CEPH_BIN/vstart-cephfs-shell << EOF
+#!/bin/bash
+
+export LD_LIBRARY_PATH=$CEPH_LIB
+export PYTHONPATH=$PYBIND:$CEPH_LIB/cython_modules/lib.$CEPH_PY_VERSION_MAJOR
```
Comment (batrick) [inline src/vstart.sh]: What I'd prefer to see here is to fix the echo'd exports above this line. I understand the difficulty is that most people don't source `vstart.sh`. Also, there's other clutter in vstart's stdout.

So perhaps let's also change the code to `tee` the environment to a file that can be sourced later, like `build/vstart_environment.sh`. Then, we can also adjust `cephfs-shell` to complain if it cannot import `rados` or `cephfs`. Maybe add some specific detection for vstart environments (i.e. does `./vstart_environment.sh` exist) so we can give a helpful error message.

File: src/vstart.sh (hunk 1 above)
Comment (jtlayton) [reply src/vstart.sh]: It doesn't seem to be. Maybe I'm missing some cmake -D magic for it?

File: src/vstart.sh (hunk 2 above)
Comment (jtlayton) [reply src/vstart.sh]: > Also, the missing executable bits on `cephfs-shell` should be backported I think. Please create a tracker for that.

What problem are we solving by backporting this? I only changed it as a convenience for running it out of the tree. The debian cephfs-shell package seems to install it with the right mode.

Comment (batrick) [reply src/vstart.sh]: Oh, then nevermind.

File: src/vstart.sh (hunk 1 above)
Comment (batrick) [reply src/vstart.sh]: It seems we just need an install directive in `src/tools/cephfs/CMakelists.txt`. @tchaikov do you remember why we didn't do that?

File: src/vstart.sh (hunk 2 above)
Comment (jtlayton) [reply src/vstart.sh]: Ok, that works. I had it just tee those values to a new vstart_environment.sh file. If you source that into your shell then cephfs-shell can be run directly.

**Comment by jtlayton** [comment]
Better version. Don't use relative paths, and ensure that we prepend values to $PYTHONPATH and $LD_LIBRARY_PATH instead of replacing them.

**Comment by jtlayton** [comment]
One more update -- add a comment header to the file that explains how to use it.

File: src/vstart.sh [hunk 3]
```diff
@@ -1224,12 +1224,17 @@ if [ "$new" -eq 1 ]; then
...
 echo ""
-echo "export PYTHONPATH=./pybind:$PYTHONPATH"
-echo "export LD_LIBRARY_PATH=$CEPH_LIB"
+echo "#" > $CEPH_DIR/vstart_environment.sh
+echo "# source this file into your shell to set up the environment." >> $CEPH_DIR/vstart_environment.sh
+echo "# For example:" >> $CEPH_DIR/vstart_environment.sh
+echo "# $ . $CEPH_DIR/vstart_environment.sh" >> $CEPH_DIR/vstart_environment.sh
+echo "#" >> $CEPH_DIR/vstart_environment.sh
+echo "export PYTHONPATH=$PYBIND:$CEPH_LIB/cython_modules/lib.${CEPH_PY_VERSION_MAJOR}:\$PYTHONPATH" | tee -a $CEPH_DIR/vstart_environment.sh
+echo "export LD_LIBRARY_PATH=$CEPH_LIB:\$LD_LIBRARY_PATH" | tee -a $CEPH_DIR/vstart_environment.sh
```
Comment (batrick) [inline src/vstart.sh]: nit:

````
{
  echo..
  echo...
} > "$CEPH_DIR/vstart_environment.sh"
````

would be cleaner.

File: src/vstart.sh [hunk 4]
```diff
@@ -1224,13 +1224,23 @@ if [ "$new" -eq 1 ]; then
     echo ""
 fi
 echo ""
-echo "export PYTHONPATH=./pybind:$PYTHONPATH"
-echo "export LD_LIBRARY_PATH=$CEPH_LIB"
+# add header to the environment file
+{
+    echo "#" > $CEPH_DIR/vstart_environment.sh
```
Comment (batrick) [inline src/vstart.sh]: You can remove the ` > $CEPH_DIR/vstart_environment.sh` from all of these echo commands now :) That was the point of the suggested change.

File: src/vstart.sh (hunk 1 above)
Comment (tchaikov) [reply src/vstart.sh]: cephfs-shell is installed using `setup.py`. so we are using `distutils_install_module()` cmake function to install it. `$CEPH_BIN` is  `${CMAKE_BINARY_DIR}/bin` which is in turn used for `CMAKE_RUNTIME_OUTPUT_DIRECTORY`. python scripts do not fall into the category of cmake "runtime target files". also, strictly speaking, we don't "install" any of files into `$CEPH_BIN`. the artifacts are installed into `${CMAKE_INSTALL_PREFIX}`. so, unless you run "make install", none of the scripts or executables will be installed into `${CMAKE_INSTALL_PREFIX}`.

File: src/vstart.sh (hunk 4 above)
Comment (jtlayton) [reply src/vstart.sh]: \<facepalm\>

**Pull request merged**